File: leetcode/cpp/rotate-image.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <vector>
using namespace std;

class Solution {
   public:
    void rotate(vector<vector<int>>& matrix) {
        vector<vector<int>>& m = matrix;
        int N = matrix.size();
        for (int r = 0; r < N / 2; r++) {
            int end = N - 1 - r;
            for (int c = r; c < end; c++) {
                int tmp = m[r][c];
                m[r][c] = m[N - 1 - c][r];
                m[N - 1 - c][r] = m[N - 1 - r][N - 1 - c];
                m[N - 1 - r][N - 1 - c] = m[c][N - 1 - r];
                m[c][N - 1 - r] = tmp;
            }
        }
    }
};

File: leetcode/rotate-image.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        m = matrix
        n = len(matrix)
        for i in range(0, n // 2):
            for j in range(i, n - 1 - i):
                # print (i, j), (j, n-1-i), (n-1-i, n-1-j), (n-1-j, i)
                v = m[n - 1 - j][i]

                m[n - 1 - j][i] = m[n - 1 - i][n - 1 - j]
                m[n - 1 - i][n - 1 - j] = m[j][n - 1 - i]
                m[j][n - 1 - i] = m[i][j]
                m[i][j] = v


if __name__ == '__main__':
    s = Solution()
    m = [[1, 2], [3, 4]]
    s.rotate(m)
    print(m)
    m = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    s.rotate(m)
    print(m)
    m = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]
    s.rotate(m)
    print(m)
